### DISCUSSION TITLE
[PHP_UNIT] Bug when assigning a false role during a test.

### DISCUSSION THREAD

**Discussion by VincentJorge** [symfony/symfony#49404] [Q&A]
### Symfony version(s) affected

5.4

### Description

to know:

Before running the unit test, I run my doctrines fixtures (which create me a user and assign him all the required roles for each route of my api).

Now I have a test that tests each of my routes in my application to make sure I get a 'forbidden' as a return to make sure my role logic still works. 

So by logic before I make my calls with my test user, I assign him a bad role.

When I browse my table containing the collection of all my routes of my application, the first route called respects the rule, so I have a forbidden return. But arrived at the second route of my table then the roles of the user were strangely refreshed, however when I make my dump:
'$user->getRoles()' during the requests I have well a single role in my return, the role 'ROLE_TO_HAVE_ACCESS_DENIED_API'. However, the following route calls are successful when they should not be, it should return a forbidden since the user does not have the right role.

### How to reproduce

(to add in a class to do a unit test via Symfony)
     
 ```
$client = self::createClient();

  $router = self::getContainer()->get('router');
  $routes = $router->getRouteCollection()->getIterator();

   $user = $this->getContainer()->get(UserRepository::class)->findOneBy(['username' => $_ENV['USER_TEST_USERNAME']]);
   
   //Actually dump(user->getRoles()); Return All roles set by fixtures load doctrine.

   $user->setRoles(['ROLE_TO_HAVE_ACCESS_DENIED_API']);
  // Now dump($user->getRoles()); return One role. 

   $token = $user->getApiToken();

foreach ($routes as $route_name => $route)
{
   $method_route = $route->getMethods()[0];
   $path_route = $route->getPath();
  $client->request(
                $method_route,
                $path_route ,
                [],
                [],
                array('HTTP_X_AUTH_TOKEN' => $token)
            );
  dd($client->getResponse()->getStatusCode()); 
  //At the first occurrence of the table the return will be a forbidden code. But then all the returns will be successes.
}
```

### Possible Solution

I think the problem is that phpunit or symfony refresh my user. 

Or maybe I'm doing it wrong and if you have a solution I'm interested.

Thank you very much

**Comment by stof** [comment]
It looks like you don't flush your change to the user after changing it. So the User is only modified in memory.

As the KernelBrowser resets the kernel between requests (to match how requests would be handled by actual requests), this memory-only change is not persistent anymore after resetting the state. Next requests will need to reload the user from the DB as the ORM identity map will have been reset, and the DB still contains your fixture roles.

**Comment by VincentJorge** [comment]
Thanks for your feedback and clarification. I'll change my logic to fit your answer.
